File: app/domain/macro.py
from uuid import uuid4


class Macro:
    
    uuid: str
    name: str
    type: str
    keys: list[str]
    
    def __init__(self, name: str, keys: list[str], uuid):
        self.name = name
        self.type = 'single' if len(keys) == 1 else 'multi'
        self.keys = keys
        if uuid is None:
            self.uuid = str(uuid4())
        else:
            self.uuid = uuid
            
    def to_dict(self):
        return {
            "uuid": self.uuid,
            "name": self.name,
            "type": self.type,
            "keys": self.keys
        }
    
    @classmethod
    def from_dict(cls, data):
        return cls(
            name=data["name"],
            keys=data["keys"],
            uuid=data["uuid"]
        )

File: kahldeck-dashboard/src/domain/Macro.ts
import { v4 as uuidv4 } from 'uuid';

export class Macro {
    readonly uuid: string;
    name: string;
    type: 'single' | 'multi';
    keys: string[];

    constructor(name: string, keys: string[], uuid?: string) {
        this.name = name;
        this.keys = keys;
        this.type = keys.length === 1 ? 'single' : 'multi';
        this.uuid = uuid || uuidv4();
    }

    toJSON(): Record<string, any> {
        return {
            uuid: this.uuid,
            name: this.name,
            type: this.type,
            keys: this.keys
        };
    }

    static fromJSON(data: Record<string, any>): Macro {
        return new Macro(
            data.name,
            data.keys,
            data.uuid
        );
    }
}